### PR TITLE
Add estimated shipment date to component builder

### PR DESCRIPTION
**Pull request by garciajordy**
# Estimated Shipment Date

This is needed when provided a property type set to
`estimatedShipmentDate`, to send the correct data to `Afterpay`.

It will first look for the variant property type, and if not found, it will look for the product property type, 
like this, you can either put it on the whole item or just the variant or even for the entire product but with one exception variant.

Co-authored-by: MassimilianoLattanzio massimiliano.lattanzio@gmail.com

### CODE REVIEW

File: app/models/solidus_afterpay/order_component_builder.rb [hunk 1]
```diff
@@ -98,8 +98,28 @@ def item(line_item)
...
+
+    def estimated_shipment_date_variant(line_item)
+      line_item.variant.variant_properties.reverse.find{ |prop| prop.property.name == "estimatedShipmentDate" }&.value
+    end
+
+    def estimated_shipment_date(line_item)
+      estimated_shipment_date_variant(line_item) || estimated_shipment_date_product(line_item)
+    end
+
+    def check_date(line_item)
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/order_component_builder.rb:119]: I think that a better name for this method could be `preorder?(line_item)`. WDYT?

File: app/models/solidus_afterpay/order_component_builder.rb [hunk 2]
```diff
@@ -98,8 +98,28 @@ def item(line_item)
...
+        estimated_shipment_date: estimated_shipment_date(line_item)
       )
     end
+
+    def estimated_shipment_date_product(line_item)
+      line_item.product.property("estimatedShipmentDate")
+    end
+
+    def estimated_shipment_date_variant(line_item)
+      line_item.variant.variant_properties.reverse.find{ |prop| prop.property.name == "estimatedShipmentDate" }&.value
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/order_component_builder.rb:112]: why did you need to call `reverse` on the variant_properties?

File: app/models/solidus_afterpay/order_component_builder.rb [hunk 3]
```diff
@@ -98,8 +98,28 @@ def item(line_item)
...
+    end
+
+    def estimated_shipment_date(line_item)
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/order_component_builder.rb:115]: because you are calling this method multiple times, wdyt about memoize the result?
```ruby
  @estimated_shipment_date ||= (estimated_shipment_date_variant(line_item) || estimated_shipment_date_product(line_item))
```

File: app/models/solidus_afterpay/order_component_builder.rb [hunk 4]
```diff
@@ -98,8 +98,28 @@ def item(line_item)
...
           currency: line_item.currency
-        )
+        ),
+        preorder: check_date(line_item),
+        estimated_shipment_date: estimated_shipment_date(line_item)
       )
     end
+
+    def estimated_shipment_date_product(line_item)
+      line_item.product.property("estimatedShipmentDate")
```
Comment (ChristianRimondi) [inline app/models/solidus_afterpay/order_component_builder.rb:108]: It's just me, but I really don't like `estimatedShipmentDate` as a property name 😄

**Review by ChristianRimondi — CHANGES_REQUESTED** [review]
I left you a couple of comments. Also, I think that we need to add few lines on the README about this new property and how it should be used.

File: app/models/solidus_afterpay/order_component_builder.rb (hunk 2 above)
Comment (garciajordy) [reply app/models/solidus_afterpay/order_component_builder.rb:112]: this was related to rubocop

File: app/models/solidus_afterpay/order_component_builder.rb (hunk 4 above)
Comment (garciajordy) [reply app/models/solidus_afterpay/order_component_builder.rb:108]: Because that's how it is on the afterpay documentation, it would be easier for people to find out to add a property called `estimatedShipmentDate` than anything else

File: app/models/solidus_afterpay/order_component_builder.rb (hunk 2 above)
Comment (ChristianRimondi) [reply app/models/solidus_afterpay/order_component_builder.rb:112]: Removing the `reverse` method doesn't trigger any rubocop offence 🤔

Comment (garciajordy) [reply app/models/solidus_afterpay/order_component_builder.rb:112]: I needed the last, not first

**Review by ChristianRimondi — APPROVED** [review]
